### PR TITLE
sync warning when async test present

### LINKED ISSUE #19: Custom validator with `async test` does not trigger invalid
This worked in yup, but does not seem to be working in jupp (at least with `validateSync`) 
Example:

```js
import * as jupp from 'jupp'

jupp.addMethod(jupp.string, 'phone', function(message: string) {
  return this.test({
    exclusive: true,
    message: message || 'phoneNumber', // message || I18n key
    name: 'phone',
    test(value: any) {
      return false  // results as invalid
    },
  })
})
```

If you change this to:

```js
jupp.addMethod(jupp.string, 'phone', function(message: string) {
  return this.test({
    exclusive: true,
    message: message || 'phoneNumber', // message || I18n key
    name: 'phone',
    async test(value: any) {
      // run a network call
      return false // results as valid - should not be
    },
  })
})
```

Then it no longer shows as invalid.

I looked through tests and found nothing for custom validations.  Any thoughts on where this might be?  Seems like a promise is not being resolved.

### PR DESCRIPTION
**Pull request by rosskevin**
Two things:
- Closes #19 and prevents users from thinking that async tests can run in sync mode
- Organizes the `sync` test by method

### CODE REVIEW

File: src/util/createValidation.js [hunk 1]
```diff
@@ -1,12 +1,28 @@
...
+  if (!sync) return Promise.resolve(result);
+
+  if (thenable(result)) {
+    throw new Error(
+      `Validation test of type: "${ctx.type}" ` +
+        'returned a Promise during a synchronous validate. ' +
+        'This test will finish after the validate call has returned',
+    );
+  }
+  return SynchronousPromise.resolve(result);
```
Comment (ctavan) [inline src/util/createValidation.js]: I would prefer to keep the `getPromise` logic here and restructure this a bit:

```js
const result = testFn.call(ctx, value);

if (sync && thenable(result)) {
  // throw
}

return getPromise(sync)....
```

**Comment by rosskevin** [comment]
@ctavan I'm not sure what you want here - promises are not my forte - perhaps it is best if we just closed this and you can add the warning on `master`?

**Comment by ctavan** [comment]
Alright, I have prepared another pull request with small amendments to your patches: #29

**Comment by rosskevin** [comment]
Great thanks!

**Pull request closed without merging**